src/entry.c

```diff
@@ -72,109 +82,123 @@ sds entryGetField(const entry *entry) {
-    field_data -= sizeof(sds);
-    return (sds *)field_data;
+    field_data -= sizeof(void *);
+    return (void **)field_data;
 }
 
-/* Returns the sds of the entry's value. */
-sds entryGetValue(const entry *entry) {
-    if (entryHasValuePtr(entry)) {
-        return *entryGetValueRef(entry);
-    } else {
+/* Returns the entry's value. */
+char *entryGetValue(const entry *entry, size_t *len) {
+    if (entryHasEmbeddedValue(entry)) {
         /* Skip field content, field null terminator and value sds8 hdr. */
         size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
-        return (char *)entry + offset;
+        sds value = (char *)entry + offset;
+        if (len) *len = sdslen(value);
+        return value;
+    }
+    void **value_ref = entryGetValueRef(entry);
+    if (!value_ref) return NULL;
+    if (entryHasStringRef(entry)) {
+        stringRef *string_ref = (stringRef *)*value_ref;
+        if (len) *len = string_ref->len;
+        return (char *)string_ref->buf;
     }
+    if (len) *len = sdslen(*value_ref);
+    return *value_ref;
 }
 
-/* Modify the value of this entry and return a pointer to the (potentially new) entry.
- * The value is taken by the function and cannot be reused after this function returns. */
-entry *entrySetValue(entry *e, sds value) {
-    if (entryHasValuePtr(e)) {
-        sds *value_ref = entryGetValueRef(e);
-        sdsfree(*value_ref);
-        *value_ref = value;
-        return e;
+static void entrySetValueSds(entry *e, sds value) {
+    serverAssert(entryHasValuePtr(e));
+    void **value_ref = entryGetValueRef(e);
+    if (entryHasStringRef(e)) {
+        zfree(*value_ref);
+        *value_ref = (stringRef *)value;
+        sdsSetAuxBit(e, FIELD_SDS_AUX_BIT_ENTRY_HAS_STRING_REF, 0);
     } else {
-        entry *new_entry = entryUpdate(e, value, entryGetExpiry(e));
-        return new_entry;
+        sdsfree(*value_ref);
     }
+    *value_ref = value;
 }
-
 /* Returns the address of the entry allocation. */
-void *entryGetAllocPtr(const entry *entry) {
+static void *entryGetAllocPtr(const entry *entry) {
     char *buf = sdsAllocPtr(entry);
-    if (entryHasValuePtr(entry)) buf -= sizeof(sds);
+    if (entryHasValuePtr(entry)) buf -= sizeof(void *);
     if (entryHasExpiry(entry)) buf -= sizeof(long long);
     return buf;
 }
 
 /**************************************** Entry Expiry API *****************************************/
+/* Returns the location of a pointer to the expiry */
+static inline long long *entryGetExpiryRef(const entry *entry) {
+    debugServerAssert(entryHasExpiry(entry));
+    char *buf = entryGetAllocPtr(entry);
+    debugServerAssert((((uintptr_t)buf & 0x7) == 0)); /* Test that the allocation is indeed 8 bytes aligned
+                                                       * This is needed since we access the expiry as with pointer casting
+                                                       * which require the access to be 8 bytes aligned. */
```

Comment:
This really isn't necessary.  Allocators are REQUIRED to align any allocation >= 8 bytes to an 8 byte boundary.  Since the entry is always more than 8 bytes, this amounts to verifying that the allocator is meeting the contract.